fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/cursors/SizeStatisticsCollectorCursor.java

```diff
@@ -0,0 +1,527 @@
+                }
+            }).whenComplete((ignore, err) -> kvCursor.close()); //no need to keep it open as it must be reconstituted with a continuation later
+        });
+    }
+
+    //update partial stats
+    private void updateStatistics(@Nonnull KeyValue kv) {
+        this.sizeStatisticsResults.keyCount += 1;
+        this.sizeStatisticsResults.keySize += kv.getKey().length;
+        this.sizeStatisticsResults.maxKeySize = Math.max(this.sizeStatisticsResults.maxKeySize, kv.getKey().length);
+        this.sizeStatisticsResults.valueSize += kv.getValue().length;
+        this.sizeStatisticsResults.maxValueSize = Math.max(this.sizeStatisticsResults.maxValueSize, kv.getValue().length);
+        int totalSize = kv.getKey().length + kv.getValue().length;
+        if (totalSize > 0) {
+            this.sizeStatisticsResults.sizeBuckets[Integer.SIZE - Integer.numberOfLeadingZeros(totalSize) - 1] += 1;
+        }
+    }
+
+    @Nonnull
+    @Override
+    public Executor getExecutor() {
+        return context.getExecutor();
+    }
+
+    @Nonnull
+    @Override
+    public boolean accept(@Nonnull RecordCursorVisitor visitor) {
+        visitor.visitEnter(this);
+        return visitor.visitLeave(this);
+    }
+
+    @Nonnull
+    private SizeStatisticsResults emitFinalStats() {
+        return sizeStatisticsResults;
+    }
+
+    @Nonnull
+    @Override
+    @Deprecated
+    public CompletableFuture<Boolean> onHasNext() {
+        throw new UnsupportedOperationException();
+    }
+
+    @Nullable
+    @Override
+    @Deprecated
+    public SizeStatisticsResults next() {
+        throw new UnsupportedOperationException();
+    }
+
+    @Nullable
+    @Override
+    @Deprecated
+    public byte[] getContinuation() {
+        throw new UnsupportedOperationException();
+    }
+
+    @Nonnull
+    @Override
+    @Deprecated
+    public RecordCursor.NoNextReason getNoNextReason() {
+        throw new UnsupportedOperationException();
+    }
+
+    @Override
+    public void close() {
+        if (hasNextFuture != null) {
+            hasNextFuture.cancel(false);
+        }
+    }
+
+    //form a continuation that allows us to restart statistics aggregation from where we left off
+    private class SizeStatisticsCollectorCursorContinuation implements RecordCursorContinuation {
+        @Nonnull
+        private final RecordCursorResult<KeyValue> currentKvResult;
+        @Nonnull
+        private final Function<byte[], RecordCursorProto.SizeStatisticsContinuation> continuationFunction;
+        @Nonnull
+        private final boolean isEnd = false;
+        @Nullable
+        private byte[] cachedBytes;
+
+        private SizeStatisticsCollectorCursorContinuation(RecordCursorResult<KeyValue> currentKvResult) {
+            this.cachedBytes = null;
+            this.currentKvResult = currentKvResult;
+
+            //defer forming bytes until requested
+            continuationFunction = b -> {
```

Comment:
Should this be a function that gets called on the class rather than a lambda (seeing as it's being set to something fixed)?

Reply:
The lambda is a closure that captures the current state needed to compute the continuation bytes at a later time. I suppose all of that state could be kept in instance variables and the toBytes method could compute the same function. It seems these are two ways to skin the same cat, or am I missing something?

Reply:
I see. Unless I'm missing something, it doesn't look like there's anything referenced from outside the closure other than the `sizeStatisticsResult` which is a class member variable. Is that right?

I think the cat gets skinned either way, but I think having a method rather than a `Function` that's always set to the same thing makes more sense from a, like, code cleanliness standpoint.

Reply:
I still think this should be a regular function rather than a `Function`.

Reply:
Scott Gray is partial to this approach. This is what was done in ConcatCursorContinuation. Perhaps we can let this one go in and then if we feel this is a wrong approach after discussing with Scott can open a Radar to change them both.

Reply:
Hm, yeah, it seems like the `ConcatCursorContinuation` should keep a member variable that is whether it is a `second` continuation rather than encapsulating it in a closure.

All right, fine, I guess.